### PR TITLE
adding hashchain detailed explanation

### PR DESCRIPTION
**Pull request by marciob**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ x ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Build (`yarn build`) was run locally and any changes were pushed
- [ ] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ x ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

There is a observation asking for more "detailed explanation" fro Hashchain, so I just added that in case it's helpful.

Issue Number: N/A

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [ x ] No

### CODE REVIEW

**Review by vivianjeng — COMMENTED** [review]
Thank you for the PR. 🙏🏻 Your description is really clear and correct.
I think the key value of hashchain is it make unirep protocol **non-repudiable**
I left some comments below.

and could you please also update the file in the path `packages/docs/docs/protocol/attestation.md`. so the next version document will be changed as well

File: packages/docs/versioned_docs/version-2.0.0-alpha-3/protocol/attestation.md [hunk 1]
```diff
@@ -12,6 +12,12 @@ UniRep offers an [`AggregateEpochKeys`](../circuits-api/circuits#aggregate-epoch
 
 ## Hashchain
 
-When an attestation is made it is stored in an on-chain mapping. When it is time to update the epoch tree root the `buildHashchain` function can be used to make a hashchain of epoch keys that need to be updated. This hashchain is output by the ZK proof and compared when it is processed.
+In the Unirep contract, a Hashchain is a data structure that is used to store a sequence of epoch keys and their corresponding reputation balances. It is used to transition user reputation between epochs in a verifiable manner. The Hashchain data structure is composed of three fields:
```
[the remark this answers is not in the document]
Comment (vimwitch) [reply packages/docs/versioned_docs/version-2.0.0-alpha-3/protocol/attestation.md:15]: Yes, technically the hashchain is an internal structure just used to update the epoch tree. The user never interacts with the hashchain, they make proofs against the resulting epoch tree.

File: packages/docs/versioned_docs/version-2.0.0-alpha-3/protocol/attestation.md [hunk 2]
```diff
@@ -12,6 +12,12 @@ UniRep offers an [`AggregateEpochKeys`](../circuits-api/circuits#aggregate-epoch
...
 
-TODO: more detailed descriptions
+* `head`: This is a reference to the first element in the Hashchain. It is used to identify the Hashchain, and is included in the zk proof generated when the Hashchain is processed.
```
[the remark this answers is not in the document]
Comment (vimwitch) [reply packages/docs/versioned_docs/version-2.0.0-alpha-3/protocol/attestation.md:17]: In the code i refer to the latest output hash of the hashchain as the head. So it's the resulting value of the hashchain.

So if there are 3 elements we have:
```
H0 = Poseidon3(attesterId, epoch, index)
H1 = Poseidon6(H0, epochKey, ...)
H2 = Poseidon6(H1, epochKey, ...)
H3 = Poseidon6(H2, epochKey, ...)
```

In this example `H3` is the `head`.

File: packages/docs/versioned_docs/version-2.0.0-alpha-3/protocol/attestation.md [hunk 3]
```diff
@@ -12,6 +12,12 @@ UniRep offers an [`AggregateEpochKeys`](../circuits-api/circuits#aggregate-epoch
...
+
+* `epochKeys`: This is an array of epoch keys that are included in the Hashchain.
+
+* `epochKeyBalances`: This is an array of reputation balances corresponding to the epoch keys in the epochKeys array.
+
+The process of transitioning user reputation between epochs involves generating a Hashchain, and submitting it to the Unirep contract for processing. The contract verifies the Hashchain by checking the provided zk proof, and then updates the user's reputation based on the data in the Hashchain. Once a Hashchain has been processed, it is marked as "processed" to prevent it from being processed again.
```
[the remark this answers is not in the document]
Comment (vimwitch) [reply packages/docs/versioned_docs/version-2.0.0-alpha-3/protocol/attestation.md:23]: I think this paragraph is confusing. We're not really talking about user state transitions on this page, it's just about attestations. I think it's better to talk about how the hashchain is used to build the latest epoch tree root, then we can mention that users can make user state transitions against this root. Maybe shorter is better?

**Comment by vimwitch** [comment]
Closing this because we're likely going to remove the hashchain structure in the next unirep pre-release. For more info see #270

**Pull request closed without merging**